src/coreclr/jit/emitarm64.cpp

```diff
@@ -5687,7 +5687,14 @@ void emitter::emitIns_R_R_I(
         // For volatile load/store, there will be memory barrier instruction before/after the load/store
         // and in such case, IsRedundantLdStr() returns false, because the method just checks for load/store
         // pair next to each other.
-        if (emitComp->opts.OptimizationEnabled() && IsRedundantLdStr(ins, reg1, reg2, imm, size, fmt))
+        if (IsRedundantLdStr(ins, reg1, reg2, imm, size, fmt))
+        {
+            return;
+        }
+
+        // If we have replaced an LDR or STR instruction with an LDP or STP then we do not want to carry on to
+        // emit the second instruction.
+        if (TryReplaceLdrStrWithPairInstr(ins, attr, reg1, reg2, imm, size, fmt))
```

Comment:
One minor way to erase some of the regression is quickly check here for `if((emitLastIns->idIns() != INS_ldr) && (emitLastIns->idIns() != INS_str))`  skip going to `TryReplaceLdrStrWithPairInstr()`. This condition is checked deeper but we could do a quick test here.

Reply:
I'm going to argue that we should not hoist any "quick checks" out of the called function, and simply hope that the native compilers when compiled with PGO/full opts will "do the right thing".

Reply:
In this case, compiler doesn't inline the calls to `IsOptimizableLdrStr` and `TryReplaceLdrStrWithPairInstr`, so we are making 2 calls before checking if the `lastIns == ins` which we can easily do upfront and skip those calls. I did `dumpbin` on `clrjit.dll` for arm64 and I see that those are not inlined.


<details>
<summary>dumpbin output of windows/arm64 clrjit.dll</summary>

```asm
?emitIns_S_R@emitter@@QEAAXW4instruction@@W4emitAttr@@W4_regNumber_enum@@HH@Z:
  0000000180007E20: A9BB53F3  stp         x19,x20,[sp,#-0x50]!
  0000000180007E24: A9015BF5  stp         x21,x22,[sp,#0x10]
  0000000180007E28: A90263F7  stp         x23,x24,[sp,#0x20]
  0000000180007E2C: A9036BF9  stp         x25,x26,[sp,#0x30]
  0000000180007E30: F90023FB  str         x27,[sp,#0x40]
  0000000180007E34: A9BE7BFD  stp         fp,lr,[sp,#-0x20]!
  0000000180007E38: 910003FD  mov         fp,sp
  0000000180007E3C: 2A0203FA  mov         w26,w2
  0000000180007E40: 290207E3  stp         w3,w1,[sp,#0x10]
  0000000180007E44: 12001749  and         w9,w26,#0x3F
  0000000180007E48: AA0003F4  mov         x20,x0
  0000000180007E4C: 2A0403F7  mov         w23,w4
  0000000180007E50: 2A0503F5  mov         w21,w5
  0000000180007E54: 52800019  mov         w25,#0
  0000000180007E58: 5280002D  mov         w13,#1
  0000000180007E5C: 5280004B  mov         w11,#2
  0000000180007E60: 7100503F  cmp         w1,#0x14
  0000000180007E64: 540002C0  beq         0000000180007EBC
  0000000180007E68: 7100543F  cmp         w1,#0x15
  0000000180007E6C: 54000240  beq         0000000180007EB4
  0000000180007E70: 7100583F  cmp         w1,#0x16
  0000000180007E74: 540001C0  beq         0000000180007EAC
  0000000180007E78: F0000A28  adrp        x8,000000018014E000
  0000000180007E7C: 9117E101  add         x1,x8,#0x5F8
  0000000180007E80: F0000A29  adrp        x9,000000018014E000
  0000000180007E84: 91190120  add         x0,x9,#0x640
  0000000180007E88: 5283D6C2  mov         w2,#0x1EB6
  0000000180007E8C: 94047487  bl          ?notYetImplemented@@YAXPEBD0I@Z
  0000000180007E90: A8C27BFD  ldp         fp,lr,[sp],#0x20
  0000000180007E94: F94023FB  ldr         x27,[sp,#0x40]
  0000000180007E98: A9436BF9  ldp         x25,x26,[sp,#0x30]
  0000000180007E9C: A94263F7  ldp         x23,x24,[sp,#0x20]
  0000000180007EA0: A9415BF5  ldp         x21,x22,[sp,#0x10]
  0000000180007EA4: A8C553F3  ldp         x19,x20,[sp],#0x50
  0000000180007EA8: D65F03C0  ret
  0000000180007EAC: 52800038  mov         w24,#1
  0000000180007EB0: 1400000B  b           0000000180007EDC
  0000000180007EB4: 52800018  mov         w24,#0
  0000000180007EB8: 14000009  b           0000000180007EDC
  0000000180007EBC: 71007C7F  cmp         w3,#0x1F
  0000000180007EC0: 54000088  bhi         0000000180007ED0
  0000000180007EC4: 7100213F  cmp         w9,#8
  0000000180007EC8: 1A8B1578  cinceq      w24,w11
  0000000180007ECC: 14000004  b           0000000180007EDC
  0000000180007ED0: 2A0903E0  mov         w0,w9
  0000000180007ED4: 940343A9  bl          ?NaturalScale_helper@emitter@@CAIW4emitAttr@@@Z
  0000000180007ED8: 2A0003F8  mov         w24,w0
  0000000180007EDC: F9400293  ldr         x19,[x20]
  0000000180007EE0: 37F80137  tbnz        x23,#0x1F,0000000180007F04
  0000000180007EE4: F9401E68  ldr         x8,[x19,#0x38]
  0000000180007EE8: D3407EE9  uxtw        x9,w23
  0000000180007EEC: D2800D0A  mov         x10,#0x68
  0000000180007EF0: 9B0A212A  madd        x10,x9,x10,x8
  0000000180007EF4: 39400149  ldrb        w9,[x10]
  0000000180007EF8: B9402948  ldr         w8,[x10,#0x28]
  0000000180007EFC: 53073936  ubfx        w22,w9,#7,#8
  0000000180007F00: 14000012  b           0000000180007F48
  0000000180007F04: F942BA6A  ldr         x10,[x19,#0x570]
  0000000180007F08: B9464669  ldr         w9,[x19,#0x644]
  0000000180007F0C: 394D8156  ldrb        w22,[x10,#0x360]
  0000000180007F10: 7100153F  cmp         w9,#5
  0000000180007F14: 54000201  bne         0000000180007F54
  0000000180007F18: 52800002  mov         w2,#0
  0000000180007F1C: 2A1703E1  mov         w1,w23
  0000000180007F20: 91020140  add         x0,x10,#0x80
  0000000180007F24: 940366D1  bl          ?tmpFindNum@RegSet@@QEBAPEAVTempDsc@@HW4TEMP_USAGE_TYPE@1@@Z
  0000000180007F28: B50000C0  cbnz        x0,0000000180007F40
  0000000180007F2C: F942BA68  ldr         x8,[x19,#0x570]
  0000000180007F30: 52800022  mov         w2,#1
  0000000180007F34: 2A1703E1  mov         w1,w23
  0000000180007F38: 91020100  add         x0,x8,#0x80
  0000000180007F3C: 940366CB  bl          ?tmpFindNum@RegSet@@QEBAPEAVTempDsc@@HW4TEMP_USAGE_TYPE@1@@Z
  0000000180007F40: B9400808  ldr         w8,[x0,#8]
  0000000180007F44: 5280002D  mov         w13,#1
  0000000180007F48: 0B150109  add         w9,w8,w21
  0000000180007F4C: 340001F6  cbz         w22,0000000180007F88
  0000000180007F50: 14000019  b           0000000180007FB4
  0000000180007F54: 350001F6  cbnz        w22,0000000180007F90
  0000000180007F58: B942D549  ldr         w9,[x10,#0x2D4]
  0000000180007F5C: 5280030C  mov         w12,#0x18
  0000000180007F60: 4B170FEA  neg         w10,w23,lsl #3
  0000000180007F64: B9408E6B  ldr         w11,[x19,#0x8C]
  0000000180007F68: 3100053F  cmn         w9,#1
  0000000180007F6C: 1A8C1128  cselne      w8,w9,w12
  0000000180007F70: 6B08015F  cmp         w10,w8
  0000000180007F74: 5400006C  bgt         0000000180007F80
  0000000180007F78: 3100053F  cmn         w9,#1
  0000000180007F7C: 1A8C112A  cselne      w10,w9,w12
  0000000180007F80: 0B0B0148  add         w8,w10,w11
  0000000180007F84: 0B150109  add         w9,w8,w21
  0000000180007F88: 528003FB  mov         w27,#0x1F
  0000000180007F8C: 1400000B  b           0000000180007FB8
  0000000180007F90: F9419D49  ldr         x9,[x10,#0x338]
  0000000180007F94: 395C1128  ldrb        w8,[x9,#0x704]
  0000000180007F98: 121F010B  and         w11,w8,#2
  0000000180007F9C: 911F3128  add         x8,x9,#0x7CC
  0000000180007FA0: 2940250A  ldp         w10,w9,[x8]
  0000000180007FA4: 4B090FE8  neg         w8,w9,lsl #3
  0000000180007FA8: 4B0B1509  sub         w9,w8,w11,lsl #5
  0000000180007FAC: 4B0A012A  sub         w10,w9,w10
  0000000180007FB0: 0B0A02A9  add         w9,w21,w10
  0000000180007FB4: 528003BB  mov         w27,#0x1D
  0000000180007FB8: F0000B28  adrp        x8,000000018016E000
  0000000180007FBC: B940F50B  ldr         w11,[x8,#0xF4]
  0000000180007FC0: 93407D36  sxtw        x22,w9
  0000000180007FC4: B5000076  cbnz        x22,0000000180007FD0
  0000000180007FC8: 52800739  mov         w25,#0x39
  0000000180007FCC: 14000023  b           0000000180008058
  0000000180007FD0: B7F801D6  tbnz        x22,#0x3F,0000000180008008
  0000000180007FD4: 1AD821A8  lsl         w8,w13,w24
  0000000180007FD8: 51000509  sub         w9,w8,#1
  0000000180007FDC: 93407D2A  sxtw        x10,w9
  0000000180007FE0: EA16015F  tst         x10,x22
  0000000180007FE4: 54000121  bne         0000000180008008
  0000000180007FE8: F10002DF  cmp         x22,#0
  0000000180007FEC: 5400036D  ble         0000000180008058
  0000000180007FF0: 9AD82AC9  asr         x9,x22,x24
  0000000180007FF4: F140053F  cmp         x9,#1,lsl #0xC
  0000000180007FF8: 5400012A  bge         000000018000801C
  0000000180007FFC: AA0903F6  mov         x22,x9
  0000000180008000: 52800759  mov         w25,#0x3A
  0000000180008004: 14000015  b           0000000180008058
  0000000180008008: 910402C8  add         x8,x22,#0x100
  000000018000800C: F107FD1F  cmp         x8,#0x1FF
  0000000180008010: 54000068  bhi         000000018000801C
  0000000180008014: 52800779  mov         w25,#0x3B
  0000000180008018: 14000010  b           0000000180008058
  000000018000801C: F9400A80  ldr         x0,[x20,#0x10]
  0000000180008020: F9405808  ldr         x8,[x0,#0xB0]
  0000000180008024: 37880108  tbnz        x8,#0x11,0000000180008044
  0000000180008028: F9402E49  ldr         x9,[xpr,#0x58]
  000000018000802C: F86B592A  ldr         x10,[x9,w11 uxtw #3]
  0000000180008030: 91400148  add         x8,x10,#0,lsl #0xC
  0000000180008034: F940150B  ldr         x11,[x8,#0x28]
  0000000180008038: B40020EB  cbz         x11,0000000180008454
  000000018000803C: 3959E568  ldrb        w8,[x11,#0x679]
  0000000180008040: 340020A8  cbz         w8,0000000180008454
  0000000180008044: AA1603E3  mov         x3,x22
  0000000180008048: 52800222  mov         w2,#0x11
  000000018000804C: 52800101  mov         w1,#8
  0000000180008050: 9401542E  bl          ?instGen_Set_Reg_To_Imm@CodeGen@@QEAAXW4emitAttr@@W4_regNumber_enum@@_JW4insFlags@@@Z
  0000000180008054: 52800819  mov         w25,#0x40
  0000000180008058: 29424FF8  ldp         w24,w19,[sp,#0x10]
  000000018000805C: 2A1903E6  mov         w6,w25
  0000000180008060: 12001745  and         w5,w26,#0x3F
  0000000180008064: AA1603E4  mov         x4,x22
  0000000180008068: 2A1B03E3  mov         w3,w27
  000000018000806C: 2A1803E2  mov         w2,w24
  0000000180008070: 2A1303E1  mov         w1,w19
  0000000180008074: AA1403E0  mov         x0,x20
  0000000180008078: 9400BD48  bl          ?IsRedundantLdStr@emitter@@AEAA_NW4instruction@@W4_regNumber_enum@@1_JW4emitAttr@@W4insFormat@1@@Z
  000000018000807C: 53001C07  uxtb        w7,w0
  0000000180008080: 35FFF087  cbnz        w7,0000000180007E90
  0000000180008084: 2A0503E6  mov         w6,w5
  0000000180008088: AA1603E5  mov         x5,x22
  000000018000808C: 2A1903E7  mov         w7,w25
  0000000180008090: 2A1B03E4  mov         w4,w27
  0000000180008094: 2A1803E3  mov         w3,w24
  0000000180008098: 2A1A03E2  mov         w2,w26
  000000018000809C: 2A1303E1  mov         w1,w19
  00000001800080A0: AA1403E0  mov         x0,x20
  00000001800080A4: 9404079B  bl          ?TryReplaceLdrStrWithPairInstr@emitter@@AEAA_NW4instruction@@W4emitAttr@@W4_regNumber_enum@@2_J1W4insFormat@1@@Z
  00000001800080A8: 53001C08  uxtb        w8,w0
  00000001800080AC: 35FFEF28  cbnz        w8,0000000180007E90
  00000001800080B0: D283FFE8  mov         x8,#0x1FFF
  00000001800080B4: EB0802DF  cmp         x22,x8
...
...
?TryReplaceLdrStrWithPairInstr@emitter@@AEAA_NW4instruction@@W4emitAttr@@W4_regNumber_enum@@2_J1W4insFormat@1@@Z:
  0000000180109F10: A9BB53F3  stp         x19,x20,[sp,#-0x50]!
  0000000180109F14: A9015BF5  stp         x21,x22,[sp,#0x10]
  0000000180109F18: A90263F7  stp         x23,x24,[sp,#0x20]
  0000000180109F1C: A9036BF9  stp         x25,x26,[sp,#0x30]
  0000000180109F20: F90023FB  str         x27,[sp,#0x40]
  0000000180109F24: A9BE7BFD  stp         fp,lr,[sp,#-0x20]!
  0000000180109F28: 910003FD  mov         fp,sp
  0000000180109F2C: D10043FF  sub         sp,sp,#0x10
  0000000180109F30: AA0003F4  mov         x20,x0
  0000000180109F34: F9400288  ldr         x8,[x20]
  0000000180109F38: 2A0603ED  mov         w13,w6
  0000000180109F3C: 2A0103F3  mov         w19,w1
  0000000180109F40: B90027ED  str         w13,[sp,#0x24]
  0000000180109F44: 2A0203F6  mov         w22,w2
  0000000180109F48: 2A0303F5  mov         w21,w3
  0000000180109F4C: 91192100  add         x0,x8,#0x648
  0000000180109F50: AA0503F9  mov         x25,x5
  0000000180109F54: 2A0703E6  mov         w6,w7
  0000000180109F58: 97FCAAE6  bl          ?OptimizationEnabled@Options@Compiler@@QEBA_NXZ
  0000000180109F5C: 53001C09  uxtb        w9,w0
  0000000180109F60: 34000A89  cbz         w9,000000018010A0B0
  0000000180109F64: AA1403E0  mov         x0,x20
  0000000180109F68: 9400005C  bl          ?emitCanPeepholeLastIns@emitter@@AEAA_NXZ
  0000000180109F6C: 53001C0C  uxtb        w12,w0
  0000000180109F70: 34000A0C  cbz         w12,000000018010A0B0
```

</details>

Reply:
In your dump the call to `?OptimizationEnabled@Options@Compiler@@QEBA_NXZ` is not even inlined, and that's a tiny function. Are you dumping a fully optimized Release build? (It's also weird the compiler does so much frame work before anything actually happens; I wonder if they didn't implement shrink wrapping on arm64)

Even without manually "partially inlining" to the caller, you could rearrange the 3 checks to check `ins` first, since it's more likely that will fail/return than "!OptimizationEnabled()" which will mostly fail/continue.

Reply:
However, if hoisting the `ins` check really does improve TP by a significant amount, maybe it is ok to hoist it, but it makes the code uglier/less maintainable.

Reply:
> Are you dumping a fully optimized Release build?

That's right, assuming our build scripts sends the right optimization flags to C++ compiler. After that I did `dumpBin` on the binary.

Reply:
Note that until this is merged we won't have PGO data for the Release build, which has some effect.

(And, our SPMI TP measurement both turns off PGO and is counting instructions, not cycles)

Reply:
> is counting instructions, not cycles

True and I am guessing if we hoist those checks, we will be executing less instructions. I think it is worth a try and see how much difference we get.

> rearrange the 3 checks to check `ins` first

You mean check with `emitLastIns` ?

Reply:
I've done some cleaning up.
* The LDP/STP check and IsRedundantLdStr() are always called one after each other. I've grouped these together inside a new inline function `OptimizeLdrStr()`.
* In the emit functions, only call `OptimizeLdrStr()` if the instruction is a `ldr` or `str`.
* Only call the LDP/STP function if the previous instruction matches the new one.
* The `OptimizationEnabled()` check is now outside of `OptimizeLdrStr()`. This isn't ideal, but putting it inside OptimizeLdrStr() gives compile errors due to emitarm64.hpp not having a full definition of compiler. Looking at the includes ordering, that seems non trivial to change.